File: 백준/Silver/1927. 최소 힙/최소 힙.java
import java.util.*;
import java.io.*;

public class Main {

    public static void main(String[] args) throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        int size = Integer.parseInt(br.readLine());
        PriorityQueue<Integer> pq = new PriorityQueue<>();

        for (int i = 0; i < size; i++) {
            int n = Integer.parseInt(br.readLine());
            if (n == 0) {
                if (pq.isEmpty()) bw.write("0\n");
                else bw.write(pq.poll() + "\n");
            } else {
                pq.offer(n);
            }
        }
        bw.flush();
        bw.close();
        br.close();
    }
}

File: 백준/Silver/1927. 최소 힙/최소 힙.kt
import java.io.*
import java.util.*

fun main(args: Array<String>) = with(BufferedReader(InputStreamReader(System.`in`))){
    val bw = BufferedWriter(OutputStreamWriter(System.out))
    val h = PriorityQueue<Int>()
    repeat(readLine().toInt()) {
        val input = readLine().toInt()
        if (input > 0) {
            h.add(input)
        } else {
            if (h.isEmpty()) bw.write("0\n")
            else bw.write("${h.remove()}\n")
        }
    }
    bw.flush()
    bw.close()
}